edb/server/smtp.py

```diff
@@ -0,0 +1,225 @@
+#
+
+from __future__ import annotations
+
+import dataclasses
+import email.message
+import asyncio
+import logging
+import os
+import hashlib
+import pickle
+import aiosmtplib
+
+from typing import Optional
+
+from edb.common import retryloop
+from edb.ir import statypes
+from edb import errors
+from . import dbview
+
+
+_semaphore: asyncio.BoundedSemaphore | None = None
+
+logger = logging.getLogger('edb.server.smtp')
+
+
+@dataclasses.dataclass
+class SMTPProviderConfig:
+    name: str
+    sender: Optional[str]
+    host: Optional[str]
+    port: Optional[int]
+    username: Optional[str]
+    password: Optional[str]
+    security: str
+    validate_certs: bool
+    timeout_per_email: statypes.Duration
+    timeout_per_attempt: statypes.Duration
+
+
+class SMTP:
+    def __init__(self, db: dbview.Database):
+        current_provider = _get_current_email_provider(db)
+        self.sender = current_provider.sender or "noreply@example.com"
+        default_port = (
+            465
+            if current_provider.security == "TLS"
+            else 587 if current_provider.security == "STARTTLS" else 25
+        )
+        use_tls: bool
+        start_tls: bool | None
+        match current_provider.security:
+            case "PlainText":
+                use_tls = False
+                start_tls = False
+
+            case "TLS":
+                use_tls = True
+                start_tls = False
+
+            case "STARTTLS":
+                use_tls = False
+                start_tls = True
+
+            case "STARTTLSOrPlainText":
+                use_tls = False
+                start_tls = None
+
+            case _:
+                raise NotImplementedError
+
+        host = current_provider.host or "localhost"
+        port = current_provider.port or default_port
+        username = current_provider.username
+        password = current_provider.password
+        validate_certs = current_provider.validate_certs
+        timeout_per_attempt = current_provider.timeout_per_attempt
+
+        req_timeout = timeout_per_attempt.to_microseconds() / 1_000_000.0
+        self.timeout_per_email = (
+            current_provider.timeout_per_email.to_microseconds() / 1_000_000.0
+        )
+        self.client = aiosmtplib.SMTP(
+            hostname=host,
+            port=port,
+            username=username,
+            password=password,
+            timeout=req_timeout,
+            use_tls=use_tls,
+            start_tls=start_tls,
+            validate_certs=validate_certs,
+        )
+
+    async def send(
+        self,
+        message: email.message.Message,
+        *,
+        test_mode: bool = False,
+    ) -> None:
+        global _semaphore
+        if _semaphore is None:
+            _semaphore = asyncio.BoundedSemaphore(
+                int(
+                    os.environ.get(
+                        "EDGEDB_SERVER_AUTH_SMTP_CONCURRENCY",
+                        os.environ.get("EDGEDB_SERVER_SMTP_CONCURRENCY", 5),
```

Comment:
wdyt?

Reply:
Can this not go through the config system somehow?

Reply:
I guess "probably, but also the existing behavior doesn't"?

Reply:
Well, now is a great time to add it to the config!